### PR TITLE
nsq levels optimization

### PR DESCRIPTION
**Pull request by NaderMahdi**
- Improve the nsq_level support by add the ability to maintain an nsq_table and reorder it based on the motion estimation information and the neighbor partition information.

- Level 1 (Up to 6 NSQ-Shapes are tested in MD) is enabled for encoder Mode 0 and it led to **~17% speed up** and **0.4 Bdrate gain**.

- Other level maybe adopted for other encoder mode in coming commits.

The description of the nsq levels is as follows:

    // NSQ search Level                               Settings
    -----------------------------------------------------------
    // NSQ_SEARCH_OFF                                 OFF
    // NSQ_SEARCH_LEVEL1                              Allow only NSQ Inter-NEAREST/NEAR/GLOBAL if parent SQ has no coeff + reordering nsq_table number and testing only 1 NSQ SHAPE
    // NSQ_SEARCH_LEVEL2                              Allow only NSQ Inter-NEAREST/NEAR/GLOBAL if parent SQ has no coeff + reordering nsq_table number and testing only 2 NSQ SHAPE
    // NSQ_SEARCH_LEVEL3                              Allow only NSQ Inter-NEAREST/NEAR/GLOBAL if parent SQ has no coeff + reordering nsq_table number and testing only 3 NSQ SHAPE
    // NSQ_SEARCH_LEVEL4                              Allow only NSQ Inter-NEAREST/NEAR/GLOBAL if parent SQ has no coeff + reordering nsq_table number and testing only 4 NSQ SHAPE
    // NSQ_SEARCH_LEVEL5                              Allow only NSQ Inter-NEAREST/NEAR/GLOBAL if parent SQ has no coeff + reordering nsq_table number and testing only 5 NSQ SHAPE
    // NSQ_SEARCH_LEVEL6                              Allow only NSQ Inter-NEAREST/NEAR/GLOBAL if parent SQ has no coeff + reordering nsq_table number and testing only 6 NSQ SHAPE
    // NSQ_SEARCH_FULL                                Allow NSQ Intra-FULL and Inter-FULL

### CODE REVIEW

**Comment by jbeich** [comment]
Appears to crash with `-enc-mode` > 2 in some cases e.g.,

```
$ curl -O https://media.xiph.org/video/derf/y4m/bus_cif.y4m
$ SvtAv1EncApp -i bus_cif.y4m -b bus_cif.ivf
-------------------------------------------
SVT-AV1 Encoder
SVT [version]:  SVT-AV1 Encoder Lib v0.4.0
SVT [build]  :  GCC 4.2.1        64 bit
LIB Build date: Mar  7 2019 00:03:43
-------------------------------------------
Number of logical cores available: 8
Number of PPCS 41
-------------------------------------------
SVT [config]: Main Profile      Tier (auto)     Level (auto)
SVT [config]: EncoderMode                                                       : 7
SVT [config]: EncoderBitDepth / CompressedTenBitFormat                          : 8 / 0
SVT [config]: SourceWidth / SourceHeight                                        : 352 / 288
SVT [config]: Fps_Numerator / Fps_Denominator / Gop Size / IntraRefreshType     : 30 / 1 / 32 / 1
SVT [config]: HierarchicalLevels / BaseLayerSwitchMode / PredStructure          : 4 / 0 / 2
SVT [config]: BRC Mode / QP  / LookaheadDistance / SceneChange                  : CQP / 50 / 33 / 0
-------------------------------------------
Encoding          

Thread 18 received signal SIGSEGV, Segmentation fault.
[Switching to LWP 101229 of process 8320]
0x00000008014d3d33 in nsq_me_analysis (p_sad8x8=0x8823be100, p_sad16x16=0x8823be0c0, p_sad32x32=0x8823be0b0,
    p_sad64x64=0x8823be0ac, p_sad64x32=0x0, p_sad32x16=0x0, p_sad16x8=0x0, p_sad32x64=0x0, p_sad16x32=0x0,
    p_sad8x16=0x0, p_sad32x8=0x0, p_sad8x32=0x0, p_sad64x16=0x0, p_sad16x64=0x0, p_nsq_64x64=0x8823bf7f8 "",
    p_nsq_32x32=0x8823bf7f9 "", p_nsq_16x16=0x8823bf7fd "", p_nsq_8x8=0x8823bf80d "")
    at Source/Lib/Codec/EbMotionEstimation.c:299
299         sad[1] = p_sad64x32[0] + p_sad64x32[1];
(gdb) backtrace full
#0  0x00000008014d3d33 in nsq_me_analysis (p_sad8x8=0x8823be100, p_sad16x16=0x8823be0c0, p_sad32x32=0x8823be0b0,
    p_sad64x64=0x8823be0ac, p_sad64x32=0x0, p_sad32x16=0x0, p_sad16x8=0x0, p_sad32x64=0x0, p_sad16x32=0x0,
    p_sad8x16=0x0, p_sad32x8=0x0, p_sad8x32=0x0, p_sad64x16=0x0, p_sad16x64=0x0, p_nsq_64x64=0x8823bf7f8 "",
    p_nsq_32x32=0x8823bf7f9 "", p_nsq_16x16=0x8823bf7fd "", p_nsq_8x8=0x8823bf80d "")
    at Source/Lib/Codec/EbMotionEstimation.c:299
        sad = {2184962556, 8, 1, 0, 3720264352}
        sad_16x8 = {162, 0, 359, 0, 307, 0, 198, 0, 162, 0, 0, 33774, 17920000, 8, 0, 27384, 2184966736, 8, 0,
          488, 1939411412, 8, 2184951232, 8, 1521128960, 8, 2186900488, 8, 2184970215, 8, 2184964228, 8}
        sad_8x16 = {8, 8, 4293394400, 1, 2184962556, 8, 2184964228, 8, 2185977865, 8, 2186900488, 8, 2184962092,
          8, 1572888, 4292935676, 4294770684, 4294901754, 4294639610, 4294901758, 4293066718, 4292935648,
          4293066718, 4292804578, 281, 0, 217, 0, 435, 0, 358, 0}
        sad_32x16 = {0, 0, 0, 0, 0, 0, 2185059092, 8}
        sad_16x32 = {5, 0, 0, 0, 0, 0, 0, 0}
        best_nsq_sad = 8
        nsq_index = 130 '\202'
        best_nsq_index = 0 '\000'
#1  0x00000008014ed784 in nsq_get_analysis_results_block (context_ptr=0x8823bb5c0)
    at Source/Lib/Codec/EbMotionEstimation.c:1440
        p_best_sad8x8 = 0x8823be100
        p_best_sad16x16 = 0x8823be0c0
        p_best_sad32x32 = 0x8823be0b0
        p_best_sad64x64 = 0x8823be0ac
        p_best_sad64x32 = 0x0
        p_best_sad32x16 = 0x0
        p_best_sad16x8 = 0x0
        p_best_sad32x64 = 0x0
        p_best_sad16x32 = 0x0
        p_best_sad8x16 = 0x0
        p_best_sad32x8 = 0x0
        p_best_sad8x32 = 0x0
        p_best_sad64x16 = 0x0
        p_best_sad16x64 = 0x0
        p_best_nsq_64x64 = 0x8823bf7f8 ""
        p_best_nsq_32x32 = 0x8823bf7f9 ""
        p_best_nsq_16x16 = 0x8823bf7fd ""
        p_best_nsq_8x8 = 0x8823bf80d ""
#2  0x00000008014e6f24 in MotionEstimateLcu (picture_control_set_ptr=0x85e52a240, sb_index=0, sb_origin_x=0,
    sb_origin_y=0, context_ptr=0x8823bb5c0, input_ptr=0x8736b7e40)
    at Source/Lib/Codec/EbMotionEstimation.c:7233
        return_error = EB_ErrorNone
        sequence_control_set_ptr = 0x85aaa9600
        xTopLeftSearchRegion = 36
        yTopLeftSearchRegion = 43
        searchRegionIndex = 21020
        picture_width = 352
        picture_height = 288
        sb_width = 64
        sb_height = 64
        padWidth = 63
        padHeight = 63
        search_area_width = 48
        search_area_height = 48
        x_search_area_origin = -32
        y_search_area_origin = -25
        origin_x = 0
        origin_y = 0
        searchRegionNumberInWidth = 0
        searchRegionNumberInHeight = 2
        xHmeLevel0SearchCenter = {{-8, -8}, {0, 84}}
        yHmeLevel0SearchCenter = {{0, 0}, {-4, 0}}
        hmeLevel0Sad = {{1850, 1850}, {2674, 2954}}
        xHmeLevel1SearchCenter = {{-8, -8}, {-8, 88}}
        yHmeLevel1SearchCenter = {{0, 0}, {0, -6}}
        hmeLevel1Sad = {{7170, 7170}, {7170, 11768}}
        xHmeLevel2SearchCenter = {{-8, -8}, {-8, 91}}
        yHmeLevel2SearchCenter = {{-1, -1}, {-1, -5}}
        hmeLevel2Sad = {{15850, 15850}, {15850, 48302}}
        xHmeSearchCenter = -8
        yHmeSearchCenter = -1
        x_search_center = -8
        y_search_center = -1
        hmeMvSad = 15850
        pu_index = 0
        max_number_of_pus_per_sb = 85
        numOfListToSearch = 1
        listIndex = 0
        candidateIndex = 0 '\000'
        totalMeCandidateIndex = 0 '\000'
        referenceObject = 0x8734a1000
        refPicPtr = 0x86d8dfb80
        quarterRefPicPtr = 0x86d8dfbe0
        sixteenthRefPicPtr = 0x86d8dfc40
        tempXHmeSearchCenter = 0
        tempYHmeSearchCenter = 0
        numQuadInWidth = 2
        totalMeQuad = 4
        quadIndex = 0
        nextQuadIndex = 0
        tempXHmeSad = 0
        ref0Poc = 0
        ref1Poc = 2
        asm_type = ASM_AVX2
        i = 0
        hmeLevel1SearchAreaInWidth = 8
        hmeLevel1SearchAreaInHeight = 8
        adjustSearchAreaDirection = 0
        enable_hme_level0_flag = 1 '\001'
        enable_hme_level1_flag = 1 '\001'
        enable_hme_level2_flag = 1 '\001'
        enableHalfPel32x32 = 1 '\001'
        enableHalfPel16x16 = 1 '\001'
        enableHalfPel8x8 = 1 '\001'
        enableQuarterPel = 1 '\001'
        oneQuadrantHME = 0 '\000'
#3  0x00000008014f845d in MotionEstimationKernel (input_ptr=0x87fd48d10)
    at Source/Lib/Codec/EbMotionEstimationProcess.c:544
        context_ptr = 0x87fd48d10
        picture_control_set_ptr = 0x85e52a240
        sequence_control_set_ptr = 0x85aaa9600
        inputResultsWrapperPtr = 0x86bb560c0
        inputResultsPtr = 0x8537ab900
        outputResultsWrapperPtr = 0x89e487500
        outputResultsPtr = 0x0
        input_picture_ptr = 0x8736b7e40
        inputPaddedPicturePtr = 0x86d8dfca0
        bufferIndex = 1936
        sb_index = 0
        xLcuIndex = 0
        yLcuIndex = 0
        picture_width_in_sb = 6
        picture_height_in_sb = 5
        sb_origin_x = 0
        sb_origin_y = 0
        sb_width = 64
        sb_height = 64
        lcuRow = 16
        paReferenceObject = 0x8734a3800
        quarterDecimatedPicturePtr = 0x86d8dfd00
        sixteenthDecimatedPicturePtr = 0x86d8dfd60
        segment_index = 0
        xSegmentIndex = 0
        ySegmentIndex = 0
        xLcuStartIndex = 0
        xLcuEndIndex = 6
        yLcuStartIndex = 0
        yLcuEndIndex = 5
        intra_sad_interval_index = 0
        asm_type = ASM_AVX2
        md_rate_estimation_array = 0x85552d870
#4  0x000000080108f776 in thread_start (curthread=0x89e487500) at /usr/src/lib/libthr/thread/thr_create.c:291
        set = {__bits = {0, 0, 0, 0}}
#5  0x0000000000000000 in ?? ()
No symbol table info available.
Backtrace stopped: Cannot access memory at address 0x7fffddbec000
```

**Comment by EwoutH** [comment]
Saw a 8,9% performance improvement (0,428 --> 0,466 fps) on a dual Xeon system (#98).

- **Video:** Morocco (1080p 8-bit 4:2:0)
- **Settings:** `-enc-mode 0 -n 120`
- **System:** 2x Intel Xeon E5-2680v4  (2x 14c/28t, AVX2), 128GB DDR4 2133MHz, Linux Mint

**Comment by jbeich** [comment]
@NaderMahdi, I confirm, 4ee345312562 fixes the above crash.

**Comment by hassount** [comment]
@EwoutH 
are you seeing a 100% CPU utilization on your system?

File: Source/Lib/Codec/EbPictureDecisionProcess.c [hunk 1]
```diff
@@ -674,17 +674,70 @@ EbErrorType signal_derivation_multi_processes_oq(
...
+    // NSQ_SEARCH_LEVEL6                              Allow only NSQ Inter-NEAREST/NEAR/GLOBAL if parent SQ has no coeff + reordering nsq_table number and testing only 6 NSQ SHAPE
+    // NSQ_SEARCH_FULL                                Allow NSQ Intra-FULL and Inter-FULL
+
+    if (MR_MODE) {
+        picture_control_set_ptr->nsq_search_level = NSQ_SEARCH_FULL;
+    }
+    else {
+        picture_control_set_ptr->nsq_search_level = NSQ_SEARCH_LEVEL6;
+    }
+
```
Comment (joelsole) [inline Source/Lib/Codec/EbPictureDecisionProcess.c]: braces not necessary

File: Source/Lib/Codec/EbProductCodingLoop.c [hunk 2]
```diff
@@ -2972,6 +2989,290 @@ void inter_depth_tx_search(
...
+    uint8_t above_size = get_part_side(above);
+    uint8_t left_size = get_part_side(left);
+    PART part = PART_N;
+
+    if (above_size == width && left_size == height) {
+        part = PART_N;
+    }
+    else if (above_size > width && left_size > height) {
+        part = PART_N;
+    }
```
Comment (joelsole) [inline Source/Lib/Codec/EbProductCodingLoop.c]: no braces according to the style guide

File: Source/Lib/Codec/EbProductCodingLoop.c [hunk 3]
```diff
@@ -2972,6 +2989,290 @@ void inter_depth_tx_search(
...
+        }
+        else {
+            context_ptr->nsq_table[5] = neighbor_part != PART_N && neighbor_part != PART_S ? neighbor_part : me_part_0;
+        }
+    }
+
+    // Remove duplicate candidates
+    for (int pidx = 0; pidx < NSQ_TAB_SIZE; pidx++) {
+        cnt[context_ptr->nsq_table[pidx]]++;
+    }
```
Comment (joelsole) [inline Source/Lib/Codec/EbProductCodingLoop.c]: braces - same below

**Comment by hassount** [comment]
@joelsole Thank you for reviewing
@NaderMahdi can you please address the comments / linux warnings ?

**Comment by hassount** [comment]
to be merged as part of #185

**Pull request closed without merging**